### PR TITLE
drop six dependency

### PR DESCRIPTION
**Pull request by jpchen**
addresses #1963 

- [x] replace methods with their py3 libraries
- [x] replace `six.reraise()` with `raise exception(..).with_traceback(...)`
- [x] replace `@add_metaclass()` with `class foo(object, metaclass=)`
- [x] fix examples

only ran the test suite, assuming we have coverage for all these functions and exception handling.

### CODE REVIEW

**Comment by jpchen** [comment]
@neerajprad  i get a precision error in
```
tests/infer/mcmc/test_hmc.py::test_unnormalized_normal[JIT=False-HMC]
AssertionError: 5.104769527853391 vs 5.0
```

related?

**Comment by neerajprad** [comment]
> related?

Not sure how your change affects that test. Please feel free to increase the threshold.

File: pyro/poutine/trace_messenger.py [hunk 1]
```diff
@@ -148,9 +146,7 @@ def __call__(self, *args, **kwargs):
             except (ValueError, RuntimeError):
                 exc_type, exc_value, traceback = sys.exc_info()
                 shapes = self.msngr.trace.format_shapes()
-                six.reraise(exc_type,
-                            exc_type(u"{}\n{}".format(exc_value, shapes)),
-                            traceback)
+                raise exc_type(u"{}\n{}".format(exc_value, shapes)).with_traceback(traceback)
```
Comment (neerajprad) [inline pyro/poutine/trace_messenger.py]: Does this get rid of the problem of duplicate stack trace that @fritzo mentioned?

File: pyro/contrib/easyguide/easyguide.py [hunk 2]
```diff
@@ -6,7 +6,6 @@
 
 import torch
 from contextlib2 import ExitStack
```
Comment (neerajprad) [inline pyro/contrib/easyguide/easyguide.py]: I think its probably time we can remove `contextlib2` from setup.py and just use the standard `contextlib` library in python 3?

Comment (fritzo) [reply pyro/contrib/easyguide/easyguide.py]: Let's add that as a task for a separate PR.

**Comment by fritzo** [comment]
I'll defer to @neerajprad for a second review.

File: pyro/contrib/easyguide/easyguide.py [hunk 3]
```diff
@@ -18,8 +17,7 @@
 from pyro.poutine.util import prune_subsample_sites
 
 
-@add_metaclass(ABCMeta)
-class EasyGuide(object):
+class EasyGuide(object, metaclass=ABCMeta):
```
Comment (neerajprad) [inline pyro/contrib/easyguide/easyguide.py]: We can also use [ABC](https://docs.python.org/3/library/abc.html#abc.ABC) as the base class directly with python 3.4+.

Comment (jpchen) [reply pyro/contrib/easyguide/easyguide.py]: Will address in follow up, I copied this pattern from the six documentation

Comment (jpchen) [reply pyro/contrib/easyguide/easyguide.py]: this syntax seems to be for python 3.6+ but i was thinking that we would support python 3.4+ WDYT?

**Pull request merged**